packages/calcite-components/src/controllers/useCancelableResource.ts

```diff
@@ -0,0 +1,74 @@
+import { makeGenericController } from "@arcgis/lumina/controllers";
+import { LitElement } from "@arcgis/lumina";
+
+/**
+ * Interface for cancelable resources.
+ */
+export interface CancelableResource {
+  cancel: () => void;
+}
+
+/**
+ * Interface for the CancelableResourceController.
+ */
+export interface UseCancelableResource {
+  /**
+   * Adds a cancelable resource to the controller.
+   *
+   * @param resource - A resource with a `cancel` method.
+   */
+  add: (resource: CancelableResource) => void;
+
+  /**
+   * Cancels all tracked resources and clears the controller.
+   */
+  cancelAll: () => void;
+
+  /**
+   * Removes a specific resource from the controller without canceling it.
+   *
+   * @param resource - The resource to remove.
+   */
+  remove: (resource: CancelableResource) => void;
+}
+
+/**
+ * A controller for managing cancelable resources.
+ *
+ * Note: resources will be canceled automatically when the component is disconnected.
+ *
+ * @param options
+ * @param options.autoCancelOnDisconnect
+ */
+export const useCancelableResource = <T extends LitElement>(options?: {
+  autoCancelOnDisconnect?: boolean;
+}): ReturnType<typeof makeGenericController<UseCancelableResource, T>> => {
+  const { autoCancelOnDisconnect = true } = options || {};
+
+  return makeGenericController<UseCancelableResource, T>((component, controller) => {
+    const args = { component, controller };
+    const { controller: adaptedController } = args;
+    const resources = new Set<CancelableResource>();
+
+    adaptedController.onDisconnected(() => {
+      if (autoCancelOnDisconnect) {
+        utils.cancelAll();
+      }
+    });
+
+    const utils: UseCancelableResource = {
+      add: (resource) => {
```

Comment:
Maybe `add` should take an array of resouces as well? Most components only have a single throttle that can be cancelled but there is definitely the possibility hat a component has multiple throttles being used and in that case it would be nicer to just call `add()` once rather than having to call `add` multiple times.